pycrdt_websocket/yroom.py

```diff
@@ -120,22 +120,25 @@ def on_message(self, value: Callable[[bytes], Awaitable[bool] | bool] | None):
         self._on_message = value
 
     async def _broadcast_updates(self):
-        if self.ystore is not None and not self.ystore.started.is_set():
-            self._task_group.start_soon(self.ystore.start)
-
-        async with self._update_receive_stream:
-            async for update in self._update_receive_stream:
-                if self._task_group.cancel_scope.cancel_called:
-                    return
-                # broadcast internal ydoc's update to all clients, that includes changes from the
-                # clients and changes from the backend (out-of-band changes)
-                for client in self.clients:
-                    self.log.debug("Sending Y update to client with endpoint: %s", client.path)
-                    message = create_update_message(update)
-                    self._task_group.start_soon(client.send, message)
-                if self.ystore:
-                    self.log.debug("Writing Y update to YStore")
-                    self._task_group.start_soon(self.ystore.write, update)
+        try:
+            if self.ystore is not None and not self.ystore.started.is_set():
+                self._task_group.start_soon(self.ystore.start)
+
+            async with self._update_receive_stream:
+                async for update in self._update_receive_stream:
+                    if self._task_group.cancel_scope.cancel_called:
+                        return
+                    # broadcast internal ydoc's update to all clients, that includes changes from the
+                    # clients and changes from the backend (out-of-band changes)
+                    for client in self.clients:
+                        self.log.debug("Sending Y update to client with endpoint: %s", client.path)
+                        message = create_update_message(update)
+                        self._task_group.start_soon(client.send, message)
```

Comment:
Should we remove a client if we cannot send them updates? Otherwise it will likely fail on every updated.
```suggestion
                    for client in self.clients:
                        try:
                            self.log.debug("Sending Y update to client with endpoint: %s", client.path)
                            message = create_update_message(update)
                            self._task_group.start_soon(client.send, message)
                        except Exception as e:
                            self.log.error("Error sending Y update to client with endpoint: %s", client.path, exc_info=e)
                            self.clients.remove(client)
```

Reply:
For send method in YDocWebsocketHandler class, it already have a try except and will capture all exception and log debug message.  https://github.com/jupyterlab/jupyter-collaboration/blob/adcde3229c986ffe498b8e6a18857195e1605ace/projects/jupyter-server-ydoc/jupyter_server_ydoc/handlers.py#L222 It is unlikely this client.send flow that raise exception.

Reply:
> also if just one send failure for temporary reason, we should probably not remove client.

Yeah, I think we need some tolerance here. Perhaps we have a time-based tolerance before removing the client. If it's not responding for some fixed time, say 1 minute, remove it. For tornado-based servers, we should likely match the [websocket ping timeout configured by jupyter_server.](https://github.com/jupyter-server/jupyter_server/blob/af1342df16963b952374b9cb110db2121423f24f/jupyter_server/serverapp.py#L255).

> But for some specific exception like WebSocketClosedError, we should remove.

As @davidbrochart already mentioned, this package is agnostic to the server framework, so we can't use this error type. However, if we add a time-based tolerance for exception, we don't have to be so specific about the exception type here.

Reply:
I think I will remove that WebsocketClosedError for now.